### PR TITLE
Qc patches

### PR DESCRIPTION
**Pull request by filmor**
This is a rollup of 6 patches from the Quantconnect fork that can be easily applied on master. With this merged, the only things separating the fork from upstream will be

1. The type conversion logic
2. A few changes to the finalizer code
3. The different versioning scheme

@AlexCatarino @Martin-Molinero @AlexCatarino, could you verify that I didn't botch your changes?

### CODE REVIEW

File: src/runtime/methodbinder.cs [hunk 1]
```diff
@@ -187,6 +187,13 @@ internal static int GetPrecedence(MethodBase mi)
                 val += ArgPrecedence(pi[i].ParameterType);
             }
 
+            var info = mi as MethodInfo;
+            if (info != null)
+            {
+                val += ArgPrecedence(info.ReturnType);
+                val += mi.DeclaringType == mi.ReflectedType ? 0 : 3000;
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs]: What does this 3000 mean? Can this be a named constant?

File: src/runtime/methodbinder.cs [hunk 2]
```diff
@@ -201,6 +208,11 @@ internal static int ArgPrecedence(Type t)
                 return 3000;
             }
 
+            if (t.IsAssignableFrom(typeof(PyObject)))
+            {
+                return -1;
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs]: That makes `TypeCode.Object` entry in the `switch` below unreachable. Is it by design?
Also, negative precedence does not sound good. What is its purpose?

File: src/runtime/propertyobject.cs [hunk 3]
```diff
@@ -160,5 +198,54 @@ public static IntPtr tp_repr(IntPtr ob)
             var self = (PropertyObject)GetManagedObject(ob);
             return Runtime.PyString_FromString($"<property '{self.info.Name}'>");
         }
+
+        private static Func<object, object> BuildGetter(PropertyInfo propertyInfo)
+        {
+            var methodInfo = propertyInfo.GetGetMethod();
+            if (methodInfo == null)
+            {
+                // if the getter is not public 'GetGetMethod' will not find it
```
Comment (lostmsu) [inline src/runtime/propertyobject.cs]: According to [MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.propertyinfo.getgetmethod?view=netframework-4.0#System_Reflection_PropertyInfo_GetGetMethod_System_Boolean_), `GetGetMethod` has an overload, that returns non-public getters, so this condition can be removed, along with `getterCacheFailed`. Same for the setters I assume

File: src/runtime/runtime.cs [hunk 4]
```diff
@@ -1741,6 +1741,11 @@ internal static bool PyObject_TypeCheck(IntPtr ob, IntPtr tp)
             return (t == tp) || PyType_IsSubtype(t, tp);
         }
 
+        internal static bool PyObjectType_TypeCheck(IntPtr type, IntPtr tp)
```
Comment (lostmsu) [inline src/runtime/runtime.cs]: For clarity, should rewrite the function above to use this function.

**Comment by lostmsu** [comment]
Also, some tests are failing apparently due to property access changes.

One test crashes during getter/setter delegate construction.

In two other cases the generated lambda converts incoming object into the type, that defined the property. For setters on value types that causes the object to be copied (via unboxing), and the original object is therefore left unmodified. Because nothing enforces in .NET requires getters to be immutable, it might affect getters too (for example, if `Lazy<T>` would be `struct`, its getter would be unable to store the created instance).

**Comment by filmor** [comment]
The tests are probably also failing due to the missing conversion changes that I'd like to discuss separately.

**Comment by filmor** [comment]
I don't see this ever getting merged, so I'll close it.

**Pull request closed without merging**